na_ucx_am_send_retry.patch

```diff
@@ -0,0 +1,171 @@
+ na_ucp_am_send_cb(
+-    void *request, ucs_status_t status, void NA_UNUSED *user_data);
++    void *request, ucs_status_t status, void *user_data);
+ 
+ /**
+  * Check if we received an AM or push the op to OP queue.
+@@ -693,6 +697,12 @@ na_ucx_complete(struct na_ucx_op_id *na_ucx_op_id, na_return_t cb_ret);
+ static NA_INLINE void
+ na_ucx_release(void *arg);
+ 
++/**
++ * Configure class parameters from environment variables.
++ */
++static na_return_t
++na_ucx_class_env_config(struct na_ucx_class * na_ucx_class);
++
+ /********************/
+ /* Plugin callbacks */
+ /********************/
+@@ -1771,14 +1781,15 @@ na_ucp_ep_close(ucp_ep_h ep)
+ /*---------------------------------------------------------------------------*/
+ static na_return_t
+ na_ucp_am_send(ucp_ep_h ep, const void *buf, size_t buf_size,
+-    const ucp_tag_t *tag, void *request)
++    const ucp_tag_t *tag, void *request, struct na_ucx_class *na_ucx_class)
+ {
+     const ucp_request_param_t send_params = {
+         .op_attr_mask = UCP_OP_ATTR_FIELD_REQUEST | UCP_OP_ATTR_FIELD_CALLBACK |
+-                        UCP_OP_ATTR_FIELD_FLAGS,
++                        UCP_OP_ATTR_FIELD_FLAGS | UCP_OP_ATTR_FIELD_USER_DATA,
+         .cb = {.send = na_ucp_am_send_cb},
+         .flags = UCP_AM_SEND_FLAG_REPLY,
+-        .request = request};
++        .request = request,
++        .user_data = na_ucx_class};
+     ucs_status_ptr_t status_ptr;
+     na_return_t ret;
+ 
+@@ -1809,9 +1820,10 @@ error:
+ 
+ /*---------------------------------------------------------------------------*/
+ static void
+-na_ucp_am_send_cb(void *request, ucs_status_t status, void NA_UNUSED *user_data)
++na_ucp_am_send_cb(void *request, ucs_status_t status, void *user_data)
+ {
+     na_return_t cb_ret;
++    struct na_ucx_class * na_ucx_class = (struct na_ucx_class *) user_data;
+ 
+     NA_LOG_SUBSYS_DEBUG(
+         msg, "ucp_am_send_nbx() completed (%s)", ucs_status_string(status));
+@@ -1820,9 +1832,13 @@ na_ucp_am_send_cb(void *request, ucs_status_t status, void NA_UNUSED *user_data)
+         NA_GOTO_DONE(done, cb_ret, NA_SUCCESS);
+     if (status == UCS_ERR_CANCELED)
+         NA_GOTO_DONE(done, cb_ret, NA_CANCELED);
++    else if (status == UCS_ERR_NO_MEMORY && na_ucx_class->remap_send_nomem)
++        NA_GOTO_SUBSYS_ERROR(msg, done, cb_ret, NA_TIMEOUT,
++            "ucp_am_send_nbx() failed (NA_NOMEM mapped to NA_TIMEOUT)");
+     else
+         NA_GOTO_SUBSYS_ERROR(msg, done, cb_ret, na_ucs_status_to_na(status),
+-            "ucp_am_send_nbx() failed (%s)", ucs_status_string(status));
++            "ucp_am_send_nbx() failed (%s), status: (%d)",
++            ucs_status_string(status), status);
+ 
+ done:
+     na_ucx_complete((struct na_ucx_op_id *) request, cb_ret);
+@@ -2022,6 +2038,9 @@ na_ucp_msg_send_cb(
+         NA_GOTO_DONE(done, cb_ret, NA_SUCCESS);
+     if (status == UCS_ERR_CANCELED)
+         NA_GOTO_DONE(done, cb_ret, NA_CANCELED);
++    else if (status = UCS_ERR_NO_MEMORY)
```

Comment:
looks like you're missing here:
```suggestion
+    else if (status = UCS_ERR_NO_MEMORY && na_ucx_class->remap_send_nomem)
```

Reply:
I think I left this incomplete (not checking the env var setting) because I inadvertently put in the remapping of out-of-memory here.  We only see it for the am_send (and not the tagged send) so I'm wondering of the remap logic belongs here (in the na_ucp_msg_send_cb).